src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -0,0 +1,2293 @@
+    loc - location in the source
+    symbol_table - symbol table
+    init_code - code block, in which declaration of some arguments
+                may be added
+
+ Purpose: Creates a string_exprt from the input exprt and adds it to
+          processed_ops
+
+\*******************************************************************/
+
+void java_string_library_preprocesst::process_single_operand(
+  exprt::operandst &processed_ops,
+  const exprt &op_to_process,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table,
+  code_blockt &init_code)
+{
+  member_exprt length(
+    op_to_process, "length", string_length_type(symbol_table));
+  member_exprt data(op_to_process, "data", string_data_type(symbol_table));
+  dereference_exprt deref_data(data, data.type().subtype());
+  string_exprt string_expr=fresh_string_expr(loc, symbol_table, init_code);
+  exprt string_expr_sym=fresh_string_expr_symbol(loc, symbol_table, init_code);
+#if 0
+  init_code.copy_to_operands(code_declt(string_expr_sym));
+#endif
+  init_code.copy_to_operands(code_assignt(string_expr.length(), length));
+  init_code.copy_to_operands(
+    code_assignt(string_expr.content(), deref_data));
+  init_code.copy_to_operands(code_assignt(string_expr_sym, string_expr));
+  processed_ops.push_back(string_expr);
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::process_operands
+
+  Inputs:
+    operands - a list of expressions
+    loc - location in the source
+    symbol_table - symbol table
+    init_code - code block, in which declaration of some arguments
+                may be added
+
+  Output: a list of expressions
+
+ Purpose: for each expression that is of a type implementing strings,
+          we declare a new `cprover_string` whose contents is deduced
+          from the expression and replace the
+          expression by this cprover_string in the output list;
+          in the other case the expression is kept as is for the output list.
+          Also does the same thing for char array references.
+
+\*******************************************************************/
+
+exprt::operandst java_string_library_preprocesst::process_operands(
+  const exprt::operandst &operands,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table,
+  code_blockt &init_code)
+{
+  exprt::operandst ops;
+  for(const auto &p : operands)
+  {
+    if(implements_java_char_sequence(p.type()))
+    {
+      dereference_exprt deref(p, to_pointer_type(p.type()).subtype());
+      process_single_operand(ops, deref, loc, symbol_table, init_code);
+    }
+    else if(is_java_char_array_pointer_type(p.type()))
+    {
+      ops.push_back(process_char_array(p, loc, symbol_table, init_code));
+    }
+    else
+    {
+      ops.push_back(p);
+    }
+  }
+  return ops;
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::process_operands_for_equals
```

Comment:
Rename process_equals_function_operands (eliminates ambiguity re: are we talking about the primitive ==, =, or .equals)